fluss-server/src/main/java/com/alibaba/fluss/server/replica/Replica.java

```diff
@@ -936,8 +936,12 @@ private boolean maybeIncrementLeaderHW(LogTablet leaderLog, long currentTimeMs)
                 leaderLog.maybeIncrementHighWatermark(newHighWatermark);
         if (oldWatermark.isPresent()) {
             LOG.debug("High watermark update from {} to {}.", oldWatermark.get(), newHighWatermark);
-            // when watermark advanced, we may need to flush kv if it's kv replica
+            // when the watermark can be advanced, we may need to flush kv first if it's kv replica,
+            // and then update highWatermark.
+            // TODO The flushKV and updateHighWatermark need to be atomic operation. See
+            // https://github.com/alibaba/fluss/issues/513
             mayFlushKv(newHighWatermark.getMessageOffset());
```

Comment:
Why not simplify move the `mayFlushKv` before `leaderLog.maybeIncrementHighWatermark`?

Reply:
Initially, I thought that the `flushKV` and `updateHighWatermark` operations should only be completed after the `newHighWatermark` is confirmed to be applied. However, in fact, the flushKV logic already includes conditions to skip actions when `newHighWatermark` is less than the current `highWatermark`.

Reply:
Done!